Test suite hits 2GB limit on 32bit R-devel

I've just seen this occur, here : 
https://ci.appveyor.com/project/Rdatatable/data-table/build/1.0.2337/job/3rppwcw31ent8g1r

Jan saw it too today here : 
https://github.com/Rdatatable/data.table/pull/2761#issuecomment-382277970

So, despite all our efforts, there's still more to do.

The error I saw is similar to Jan's : 
```
  Running test id 1667.1      Error: memory exhausted (limit reached?)
  Error: memory exhausted (limit reached?)
  Error: memory exhausted (limit reached?)
  Execution halted
```

That message is R's, here : 
https://github.com/wch/r-source/blob/trunk/src/main/memory.c#L2059
and looking at the context there (inside an err_malloc function) suggests AppVeyor instance is declining R's malloc request.  (i.e. it doesn't seem like some kind of stack limit inside R itself).

@jangorecki 
* What is the memory limit of an AppVeyor instance?  I had a quick look and 4GB came up quickly but I'm not sure if that's correct.  If it is 4GB I don't see how we are reaching that.
* We never knew before, but now with your new memtest charts, it seems the max memory used is towards the end of the test suite and reaches 160MB.  If that's right, it seems very low.  I can't imagine AppVeyor's limit being that low.  Certainly a lot less than 4GB and we know it now passes on your 2GB instance.
* I already reduced the memory usage of the non-equi-join tests in PR #2733.  I'm fairly confident that helped a lot and that the memory usage is no longer in that area.
* According to the memtest charts there are 3 large step-ups in memory usage very early in the test suite. If the mem usage table could be sorted by change and the top 10 printed at the end of the test suite then we could see which tests those step-ups are for.  If the data used in those 3 tests is `rm()`'d after those tests, then perhaps more memory will be available for the tests at the end.   However, as I said above, it's strange that the memtest charts show only 160MB being used!  Reducing those 3 tests will be wasted effort if, somehow, the memtest results are not capturing the actual mem usage on AppVeyor.
* A long shot ... might it be that a Suggests package test is running on AppVeyor which eats more RAM, which does not run in the environment used to create the memtest charts.

Is this installation error for `zlibbioc` (a dependency of `GenomicRanges`) illuminating?

```
trying URL 'https://bioconductor.org/packages/release/bioc/src/contrib/zlibbioc_1.24.0.tar.gz'
Content type 'application/x-gzip' length 257311 bytes (251 KB)
downloaded 251 KB

* installing *source* package 'zlibbioc' ...

   **********************************************
   WARNING: this package has a configure script
         It probably needs manual configuration
   **********************************************


** libs

*** arch - i386
mkdir -p "C:/R/R-3.5.0/library/zlibbioc/include" "C:/R/R-3.5.0/library/zlibbioc/libs/i386"
make -C zlib-1.2.5 -fwin32/Makefile.gcc libzbioc.a zlib1bioc.dll
make[1]: Entering directory `/cygdrive/c/Users/hughp/AppData/Local/Temp/Rtmp6HboKw/R.INSTALLb0c7c2c5bc/zlibbioc/src-i386/zlib-1.2.5'
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o adler32.o adler32.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o compress.o compress.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o crc32.o crc32.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o deflate.o deflate.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o gzclose.o gzclose.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o gzlib.o gzlib.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o gzread.o gzread.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o gzwrite.o gzwrite.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o infback.o infback.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o inffast.o inffast.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o inflate.o inflate.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o inftrees.o inftrees.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o trees.o trees.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o uncompr.o uncompr.c
gcc -m32  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o zutil.o zutil.c
ar rcs libzbioc.a adler32.o compress.o crc32.o deflate.o gzclose.o gzlib.o gzread.o gzwrite.o infback.o inffast.o inflate.o inftrees.o trees.o uncompr.o zutil.o 
windres --define GCC_WINDRES --target=pe-i386 -o zlibrc.o win32/zlib1.rc
gcc -m32 -shared -Wl,--out-implib,libzbiocdll.a  \
-o zlib1bioc.dll win32/zlib.def adler32.o compress.o crc32.o deflate.o gzclose.o gzlib.o gzread.o gzwrite.o infback.o inffast.o inflate.o inftrees.o trees.o uncompr.o zutil.o  zlibrc.o
strip zlib1bioc.dll
make[1]: Leaving directory `/cygdrive/c/Users/hughp/AppData/Local/Temp/Rtmp6HboKw/R.INSTALLb0c7c2c5bc/zlibbioc/src-i386/zlib-1.2.5'
cp zlib-1.2.5/libzbioc.a zlib-1.2.5/zlib1bioc.dll \
            "C:/R/R-3.5.0/library/zlibbioc/libs/i386"
cp zlib-1.2.5/zconf.h zlib-1.2.5/zlib.h "C:/R/R-3.5.0/library/zlibbioc/include"
c:/Rtools/mingw_32/bin/gcc  -I"C:/R/R-35~1.0/include" -DNDEBUG          -O3 -Wall  -std=gnu99 -mtune=generic -c R_init_zlibbioc.c -o R_init_zlibbioc.o
c:/Rtools/mingw_32/bin/gcc -shared -s -static-libgcc -o zlibbioc.dll tmp.def R_init_zlibbioc.o -LC:/R/R-35~1.0/bin/i386 -lR
installing to C:/R/R-3.5.0/library/zlibbioc/libs/i386

*** arch - x64
mkdir -p "C:/R/R-3.5.0/library/zlibbioc/include" "C:/R/R-3.5.0/library/zlibbioc/libs/x64"
make -C zlib-1.2.5 -fwin32/Makefile.gcc libzbioc.a zlib1bioc.dll
make[1]: Entering directory `/cygdrive/c/Users/hughp/AppData/Local/Temp/Rtmp6HboKw/R.INSTALLb0c7c2c5bc/zlibbioc/src-x64/zlib-1.2.5'
gcc -m64  -O3 -Wall -DNO_VIZ -DZ_PREFIX -c -o adler32.o adler32.c
adler32.c:1:0: sorry, unimplemented: 64-bit mode not compiled in
 /* adler32.c -- compute the Adler-32 checksum of a data stream
 ^
make[1]: *** [adler32.o] Error 1
make[1]: Leaving directory `/cygdrive/c/Users/hughp/AppData/Local/Temp/Rtmp6HboKw/R.INSTALLb0c7c2c5bc/zlibbioc/src-x64/zlib-1.2.5'
make: *** [libzbioc] Error 2
ERROR: compilation failed for package 'zlibbioc'
* removing 'C:/R/R-3.5.0/library/zlibbioc'
In R CMD INSTALL
Warning in install.packages :
  installation of package ‘C:/Users/hughp/AppData/Local/Temp/Rtmpy6Q6nh/downloaded_packages/zlibbioc_1.24.0.tar.gz’ had non-zero exit status
```

--- a/CRAN_Release.cmd
+++ b/CRAN_Release.cmd
@@ -95,16 +95,15 @@ grep ScalarString *.c
 # If a PROTECT is not needed then a comment is added explaining why and including "PROTECT" in the comment to pass this grep
 grep allocVector *.c | grep -v PROTECT | grep -v SET_VECTOR_ELT | grep -v setAttrib | grep -v return
 
-cd
 R
 cc(clean=TRUE)  # to compile with -pedandic. Also use very latest gcc (currently gcc-7) as CRAN does
 saf = options()$stringsAsFactors
 options(stringsAsFactors=!saf)    # check tests (that might be run by user) are insensitive to option, #2718
-cc()
+test.data.table()
 q("no")
-R CMD build data.table
-R CMD check data.table_1.10.1.tar.gz --as-cran    # remove.packages("xml2") first to prevent the 150 URLs in NEWS.md being pinged by --as-cran
-R CMD INSTALL data.table_1.10.1.tar.gz
+R CMD build .
+R CMD check data.table_1.10.5.tar.gz --as-cran    # remove.packages("xml2") first to prevent the 150 URLs in NEWS.md being pinged by --as-cran
+R CMD INSTALL data.table_1.10.5.tar.gz
 R
 require(data.table)
 test.data.table()
@@ -128,6 +127,7 @@ make
 alias R310=~/build/R-3.1.0/bin/R
 ### END ONE TIME BUILD
 
+cd ~/GitHub/data.table
 R310 CMD INSTALL ./data.table_1.10.5.tar.gz
 R310
 require(data.table)
@@ -199,18 +199,20 @@ sudo apt-get purge libcurl4-openssl-dev    # cannot coexist, it seems
 sudo apt-get install libcurl4-openssl-dev:i386
 cd ~/build/32bit/R-devel
 ./configure --without-recommended-packages --disable-byte-compiled-packages --disable-openmp --without-readline --without-x CC="gcc -m32" CXX="g++ -m32" F77="gfortran -m32" FC=${F77} OBJC=${CC} LDFLAGS="-L/usr/local/lib" LIBnn=lib LIBS="-lpthread" CFLAGS="-O0 -g -Wall -pedantic"
+##
 
 make
 alias Rdevel='~/build/R-devel/bin/R --vanilla'
+cd ~/GitHub/data.table
 Rdevel CMD INSTALL data.table_1.10.5.tar.gz
 # Check UBSAN and ASAN flags appear in compiler output above. Rdevel was compiled with
 # them so should be passed through to here
 Rdevel
-install.packages("bit64")  # important to run tests using integer64
+install.packages(c("bit64","xts","nanotime","chron"), repos="http://cloud.r-project.org")  # important to run all tests
 # Skip compatibility tests with other Suggests packages; unlikely UBSAN/ASAN problems there.
 require(data.table)
-require(bit64)
-test.data.table()     # slower than usual of course due to UBSAN and ASAN. Too slow to run R CMD check.
+test.data.table()     # slower than usual, naturally, due to UBSAN and ASAN. Too slow to run R CMD check.
+for (i in 1:10) test.data.table()  # last resort: try several runs; e.g a few tests generate data with a non-fixed random seed
 # Throws /0 errors on R's summary.c (tests 648 and 1185.2) but ignore those: https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=16000
 q("no")
 

--- a/src/ijoin.c
+++ b/src/ijoin.c
@@ -145,7 +145,9 @@ SEXP lookup(SEXP ux, SEXP xlen, SEXP indices, SEXP gaps, SEXP overlaps, SEXP mul
         if (!len1[i]) continue;
         vv = VECTOR_ELT(lookup, i);
         tt = VECTOR_ELT(type_lookup, i);
-        INTEGER(tt)[0] = INTEGER(vv)[0];
+        if (length(tt) && length(vv)) {            // length check added by Matt to avoid SEGV in #2767
+          INTEGER(tt)[0] = INTEGER(vv)[0];
+        }
       }
       break;
 
@@ -154,7 +156,9 @@ SEXP lookup(SEXP ux, SEXP xlen, SEXP indices, SEXP gaps, SEXP overlaps, SEXP mul
         if (!len1[i]) continue;
         vv = VECTOR_ELT(lookup, i);
         tt = VECTOR_ELT(type_lookup, i);
-        INTEGER(tt)[0] = INTEGER(vv)[len1[i]-1];
+        if (length(tt) && length(vv)>=len1[i]) {   // length check added by Matt to avoid SEGV in #2767
+          INTEGER(tt)[0] = INTEGER(vv)[len1[i]-1];
+        }
       }
 
     case ALL :

